px4-workspace/pictorus-px4/src/ffi_protocol.rs

```diff
@@ -0,0 +1,1132 @@
+//!
+//! The idea is to provide a way for pictorus static libs to to communicate larger amounts of data than
+//! what we currently pass through FFI variables. This Rust code requires C/C++ code on the calling side
+//! which is using the exact same memory layout that the rust code expects.
+use crate::message_impls::{Topic, UorbMessage};
+use once_cell::sync::Lazy;
+use pictorus_traits::{InputBlock, OutputBlock, Pass, PassBy};
+use px4_msgs_sys::orb::orb_id_t;
+use spin::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+extern crate alloc;
+use alloc::{boxed::Box, vec, vec::Vec};
+
+use crate::message_impls::{FromPassType, ToPassType};
+
+/// C-compatible error codes for FFI boundary
+#[repr(C)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FfiReturnCode {
+    /// Success - no error
+    Success = 0,
+    /// Message length mismatch
+    MessageLengthMismatch = 1,
+    /// Attempt to get message type that has not been advertised
+    UnadvertisedMessage = 2,
+    /// Attempt to get message type that has not been subscribed
+    UnsubscribedMessage = 3,
+    /// Invalid message index
+    InvalidMessageIndex = 4,
+    /// Null argument(s) passed to function
+    NullArgument = 5,
+}
+
+impl FfiReturnCode {
+    /// Returns true if this represents a success state
+    pub fn is_success(self) -> bool {
+        self == FfiReturnCode::Success
+    }
+
+    /// Returns true if this represents an error state
+    pub fn is_error(self) -> bool {
+        !self.is_success()
+    }
+}
+
+/// Core FFI protocol manager for PX4-Rust message exchange
+///
+/// This struct manages the bidirectional message passing system between PX4 C++ modules
+/// and Rust computation code. It maintains separate collections of input and output messages,
+/// handling memory management and synchronization safely across the FFI boundary.
+///
+/// # Thread Safety
+///
+/// The protocol is designed for single-threaded operation within PX4 modules. While it uses
+/// `RwLock` for static requirements, there should be no lock contention in normal operation.
+///
+/// # Memory Management
+///
+/// All message data is stored in owned heap buffers (`Box<[u8]>`) to ensure memory safety
+/// and proper cleanup. The protocol handles the complexity of C-compatible memory layout
+/// while providing Rust safety guarantees.
+///
+/// # Examples
+///
+/// ```rust
+/// use pictorus_px4::ffi_protocol::FfiProtocol;
+/// use pictorus_px4::message_impls::SensorAccel;
+///
+/// // Get access to the global protocol instance
+/// let mut protocol = FfiProtocol::get_mut();
+///
+/// // Set up message subscriptions
+/// protocol.subscribe_to_message(SensorAccel::default());
+/// ```
+pub struct FfiProtocol {
+    /// Input messages that C++ writes and Rust reads
+    input_messages: Vec<MessageEntry>,
+    /// Output messages that Rust writes and C++ reads  
+    output_messages: Vec<MessageEntry>,
+}
+
```

Comment:
I have vague dreams of using an intrusive linked list where each instance of an FfiOutputBlock/FfiInputBlock has statically allocated memory for these. This would allow us to drop the `alloc` requirement. Definitely low priority but I just wanted to put that idea down for posterity.